techniques/system/common/1.0/update.st

```diff
@@ -143,13 +143,64 @@ bundle agent update_action
   vars:
 
       "client_inputs"        string => "${sys.workdir}/inputs";  #where to put the files on the client when downloaded
-      "file_to_check_update" string => "rudder_promises_generated";
-    nova_edition::
-      "server_inputs"        string => "${server_info.policy_files}/rules/cfengine-nova"; #actual directory with promises
-    community_edition::
-      "server_inputs"        string => "${server_info.policy_files}/rules/cfengine-community"; #actual directory with promises
+
+       # The check is in two phase:
+       # First copy rudder_promises_timestamp_tmp from the policy server, and check is it's newer
+       # (based on the timestamp written in it) than rudder_promises_timestamp.
+       # If so, copy all the promises
+       "rudder_promises_timestamp"     string => "rudder_promises_generated";
+       "rudder_promises_timestamp_tmp" string => "rudder_promises_generated.tmp";
+
+       # Same for rudder-tools
+       "rudder_tools_updated_tmp"   string => "${g.rudder_tools_updated_filename}.tmp";
+
+     current_promise_timestamp_defined::
+       "current_promises_timestamp" string => readfile("${client_inputs}/${rudder_promises_timestamp}", 2048);
+     current_tools_timestamp_defined::
+       "current_tools_timestamp"   string => readfile("${g.rudder_tools_updated}", 2048);
+
+     # File ought to be there only in second pass, but since classes are persistent, we need to 
+     # enforce the check on pass_2
+     pass2.(rudder_promises_generated_tmp_file_kept|rudder_promises_generated_tmp_file_repaired)::
```

Comment:
This is an excellent question.
Classes are persistent, and error classes are canceled in case of success; so it's probably unlikely to have error and repaired at the same time.
However, to be on the safe side; i'll add the .!error condition